File: ios/Classes/PntaFlutterPlugin.swift
import Flutter
import UIKit
import UserNotifications


public class PntaFlutterPlugin: NSObject, FlutterPlugin, UIApplicationDelegate {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "pnta_flutter", binaryMessenger: registrar.messenger())
    let instance = PntaFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.addApplicationDelegate(instance)
    ForegroundNotificationHandler.register(with: registrar)
    NotificationTapHandler.register(with: registrar)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "requestNotificationPermission":
      PermissionHandler.requestNotificationPermission(result: result)
    case "checkNotificationPermission":
      PermissionHandler.checkNotificationPermission(result: result)
    case "getDeviceToken":
      TokenHandler.getDeviceToken(result: result)
    case "identify":
      if let args = call.arguments as? [String: Any],
         let projectId = args["projectId"] as? String {
        let pntaSdkVersion = args["pntaSdkVersion"] as? String ?? "Unknown"
        IdentifyHandler.identify(projectId: projectId, pntaSdkVersion: pntaSdkVersion, result: result)
      } else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing arguments for identify", details: nil))
      }
    case "updateMetadata":
      if let args = call.arguments as? [String: Any],
         let projectId = args["projectId"] as? String {
        let metadata = args["metadata"] as? [String: Any]
        MetadataHandler.updateMetadata(projectId: projectId, metadata: metadata, result: result)
      } else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing arguments for updateMetadata", details: nil))
      }
    case "setForegroundPresentationOptions":
      if let args = call.arguments as? [String: Any],
         let showSystemUI = args["showSystemUI"] as? Bool {
        ForegroundNotificationHandler.setForegroundPresentationOptions(showSystemUI: showSystemUI)
        result(nil)
      } else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing showSystemUI argument", details: nil))
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // UIApplicationDelegate methods for APNs token forwarding
  public func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
    TokenHandler.didRegisterForRemoteNotifications(deviceToken: deviceToken)
  }

  public func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
    TokenHandler.didFailToRegisterForRemoteNotifications(error: error)
  }
}

File: android/src/main/kotlin/io/pnta/pnta_flutter/PntaFlutterPlugin.kt
package io.pnta.pnta_flutter

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import android.app.Activity
import io.pnta.pnta_flutter.PermissionHandler
import io.pnta.pnta_flutter.TokenHandler
import io.pnta.pnta_flutter.IdentifyHandler
import io.pnta.pnta_flutter.ForegroundNotificationHandler
import io.pnta.pnta_flutter.NotificationTapHandler
import android.app.NotificationChannel
import android.app.NotificationManager
import android.os.Build
import android.content.Context
import io.flutter.plugin.common.PluginRegistry
import android.content.Intent

/** PntaFlutterPlugin */
class PntaFlutterPlugin: FlutterPlugin, MethodCallHandler, ActivityAware, PluginRegistry.NewIntentListener {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private var activity: Activity? = null
  private var pluginBinding: ActivityPluginBinding? = null

  private fun createDefaultNotificationChannel(context: Context) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      val channelId = "pnta_default"
      val channelName = "General Notifications"
      val importance = NotificationManager.IMPORTANCE_DEFAULT
      val channel = NotificationChannel(channelId, channelName, importance)
      val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
      notificationManager.createNotificationChannel(channel)
    }
  }

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "pnta_flutter")
    channel.setMethodCallHandler(this)
    
    ForegroundNotificationHandler.register(flutterPluginBinding.binaryMessenger)
    NotificationTapHandler.register(flutterPluginBinding.binaryMessenger)
    // Create default notification channel on engine attach
    createDefaultNotificationChannel(flutterPluginBinding.applicationContext)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    if (call.method == "requestNotificationPermission") {
      PermissionHandler.requestNotificationPermission(activity, result)
    } else if (call.method == "getDeviceToken") {
      TokenHandler.getDeviceToken(activity, result)
    } else if (call.method == "identify") {
      val projectId = call.argument<String>("projectId")
      val metadata = call.argument<Map<String, Any>>("metadata")
      val pntaSdkVersion = call.argument<String>("pntaSdkVersion") ?: "Unknown"
      if (projectId == null) {
        result.error("INVALID_ARGUMENTS", "projectId is null", null)
        return
      }
      IdentifyHandler.identify(activity, projectId, metadata, pntaSdkVersion, result)
    } else if (call.method == "updateMetadata") {
      val projectId = call.argument<String>("projectId")
      val metadata = call.argument<Map<String, Any>>("metadata")
      if (projectId == null) {
        result.error("INVALID_ARGUMENTS", "projectId is null", null)
        return
      }
      MetadataHandler.updateMetadata(projectId, metadata, result)
    } else if (call.method == "setForegroundPresentationOptions") {
      val showSystemUI = call.argument<Boolean>("showSystemUI") ?: false
      ForegroundNotificationHandler.setForegroundPresentationOptions(showSystemUI)
      result.success(null)
    } else {
      result.notImplemented()
    }
  }

  fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray): Boolean {
    return PermissionHandler.onRequestPermissionsResult(requestCode, permissions, grantResults)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  // ActivityAware methods
  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
    pluginBinding = binding
    binding.addRequestPermissionsResultListener { requestCode, permissions, grantResults ->
      onRequestPermissionsResult(requestCode, permissions, grantResults)
    }
    binding.addOnNewIntentListener(this)
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
    pluginBinding?.removeOnNewIntentListener(this)
    pluginBinding = null
    PermissionHandler.cleanup()
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
    pluginBinding = binding
    binding.addOnNewIntentListener(this)
  }

  override fun onDetachedFromActivity() {
    activity = null
    pluginBinding?.removeOnNewIntentListener(this)
    pluginBinding = null
    PermissionHandler.cleanup()
  }

  override fun onNewIntent(intent: Intent): Boolean {
    val extras = intent.extras
    if (extras != null && !extras.isEmpty) {
      val payload = mutableMapOf<String, Any>()
      for (key in extras.keySet()) {
        val value = extras.get(key)
        when (value) {
          is String, is Int, is Boolean, is Double, is Float, is Long -> payload[key] = value
          else -> if (value != null) payload[key] = value.toString()
        }
      }
      if (payload.isNotEmpty()) {
        NotificationTapHandler.sendTapPayload(payload)
      }
    }
    return false // allow other listeners to process
  }
}
